### PR TITLE
More precise completions and namespace member completions

### PR DESCRIPTION
**Pull request by minestarks**
## Feature overview

This PR enables:

1.  Much more precise keyword completions where we only show the relevant syntax at a particular cursor location. For example:

```qsharp
within { } | 
```

At the cursor location indicated by `|`, we should only get the keyword `apply` .

2. More precise completions for names where we take the syntactic context into account. For example:

```qsharp
operation Foo(x : | )
```

Here, we should only get type names.

3. Namespace member completions. For example:
```qsharp
import Std.Diagnostics.|
```

The completions here should return only items that exist under `Std.Diagnostics`.

### Parser completions module

Introduce the `completion` module in the parser.

- `possible_words_at_offset_in_source` - a new API to return completions at a particular offset . Uses the parser directly as the best source of knowledge about what tokens to expect.
- Introduce the `WordKinds` flags to define the different kinds of word tokens that the parser can expect during parsing. (See `compiler/qsc_parse/src/completion/word_kinds.rs` for longer description)
- See `compiler/qsc_parse/src/completion/collector.rs` for a description of how this all works.
- Sprinkle `expect()`s at all possible locations in the parser where we expect a word.

### Error recovery - AST changes

- Introduce `PathKind::Ok` and `PathKind::Err` .
    - This new approach allows for a valid AST node to exist even when path parsing fails for a partial path. This is good for syntax such as "`open Std.`" where we need:
        - The identifier nodes preceding `.` to be valid, so we can use them to look up possible completions.
        - The `open` statement to also be a valid node, so we can tell that the context requires a namespace (and not, say, a type name).
    - Most existing code will continue to only process valid paths. 
- Remove `Idents` struct and replace it with a trait implemented by various types such as `Path` and `Ident` slices. This should make the convenience methods easier to use without cloning in/out of the different types we need.

### Error recovery - parser changes

- Recovering struct parsing, so that e.g.  `new Foo.`  is still parsed as a valid struct expression node, allowing us to access the `Foo` ident from the AST
- Recovering `open` statement parsing so that `open Foo.` can be recognized as an open item in the AST.

### Language service

In `language_service` is essentially a complete rewrite of `completions.rs`, so I suggest reviewing the new code only and ditching the diff.

### Other parser changes

- Move implicit namespace parsing a bit so that it works better with the completion  mechanism.
    - This also fixes the  spurious "expected item, found eof" error when document only contains one token
- Delete the dead path  `path_field_op` 
- Add some missing keywords to item parsing recovery (`export`, `import`)

### Other notes

- Locals completions exclude `import`s from locals (they conflict with globals already added to completion) (fix in `resolve.rs`)
- `language-configuration.json` : and removed wonky word pattern in playground and VS Code because we don't need it anymore now that we have proper `.` completions! Also update trigger characters.
- Only include the samples snippets in empty documents, because now with the more concise completion list, it gets quite intrusive for them to show up at every word when typing code.
- Update integration test to account for the fact that completions are more accurate now.

### Not in this PR (coming next)

- Struct and UDT field access completions.

### CODE REVIEW

File: compiler/qsc_ast/src/ast.rs [hunk 1]
```diff
@@ -1947,36 +1930,24 @@ impl Display for ImportOrExportItem {
...
-            is_glob: self.is_glob,
-        }
+        Self { span, ..self }
     }
 }
 
 impl ImportOrExportItem {
-    /// Returns the span of the export item. This includes the path and , if any exists, the alias.
+    /// Returns the alias ident, if any, or the name from the path if no alias is present.
+    /// Returns `None` if this the path has an error.
```
Comment (swernli) [inline compiler/qsc_ast/src/ast.rs:1939]: nitpick: extra word

```suggestion
    /// Returns `None` if the path has an error.
```

**Comment by minestarks** [comment]
> _Change in memory usage detected by benchmark._
> 
> ## Memory Report for [6fbbec5](https://github.com/microsoft/qsharp/commit/6fbbec5027789cfc5e384b13e4eb61f410f50b52)
> Test	This Branch	On Main	Difference
> compile core + standard lib	17940700 bytes	17932132 bytes	8568 bytes

I ran some tests to narrow it down. This increase is attributable entirely to the changes in the AST. A bunch of `Path` or `Idents` fields have been replaced with `PathKind`, a `Span` field has been added to `ImportOrExportItem`. It adds up to a net increase in memory.

I can't think of a way to collapse the AST any further, we are storing more information in there after all. So I think the 0.04% increase in memory usage is a worthwhile tradeoff for the feature.

File: language_service/src/completion.rs [hunk 2]
```diff
@@ -1,809 +1,334 @@
...
+        CompilationKind::OpenProject {
+            package_graph_sources,
+        } => possible_words_at_offset_in_source(
+            source_contents,
+            Some(source_name_relative),
+            package_graph_sources.root.language_features,
+            cursor_offset,
+        ),
+        CompilationKind::Notebook { project } => possible_words_at_offset_in_fragments(
+            source_contents,
```
Comment (swernli) [inline language_service/src/completion.rs:111]: How do source contents work for notebook cells? I was curious about this and tried playing around with it, ended up seeing some behaviors I didn't expect, so wanted to confirm whether there are known bugs/limitations here or if this is something that bears investigating. I'll follow up with screen shots and more explanation...

Comment (swernli) [reply language_service/src/completion.rs:111]: When I try out the language service with these changes, completions in a file work as I'd expect. I get completions for local variables that are in-scope at my current line (the `x` from the line above):
image
or for locally defined callables:
image

But in notebook cells, I don't see these elements if they come from top-level statements (it does work as expected when inside a callable context). So the local variables don't show `x` or `IntOne` from the lines above my cursor:
image
image

Oddly enough, the locals and functions DO show up under two different circumstances. If instead of triggering completions with ctrl+space I just start typing, they show up:
image
image
OR if I put the definitions in a previous cell, then they show up even if I trigger completions before I start typing:
image

I don't think this is a major issue and would probably be fine as a follow up investigation/bug-fix, but it does tell me that there is some difference between the response of completions when triggered manually vs when typing specifically for notebook cells that I don't really understand.

File: language_service/src/completion/global_items.rs
Comment (minestarks) [inline language_service/src/completion/global_items.rs:1]: Hmm, just discovered this file will have to undergo some rewrites. I'd suggest skipping most of it for now. The `pub` structs / methods should stay the same though, so the rest of the PR should be uneffected.

@swernli FYI since I know you've started looking.

After talking to @sezna about namespaces and exports , there's a much simpler way to gather the names under a particular namespace than what I did here. I'll try that approach and post an update here when I'm done.

File: language_service/src/tests.rs [hunk 3]
```diff
@@ -210,18 +210,19 @@ async fn completions_requested_after_document_load() {
...
                 line: 0,
-                column: 76
-            }
+                column: 92,
+            },
         )
         .items
         .len(),
-        13
     );
```
Comment (idavis) [inline language_service/src/tests.rs:226]: Would it aid debugging test failures to compare the lists instead of the number of items in the lists?

File: playground/src/main.tsx [hunk 4]
```diff
@@ -310,7 +304,7 @@ function registerMonacoLanguageServiceProviders(
         }),
       };
     },
-    triggerCharacters: ["@"], // for attribute completions
+    triggerCharacters: ["@", "."],
```
Comment (idavis) [inline playground/src/main.tsx:307]: Should we add a note that we need to keep this list in sync with the one in `extension.ts`?

File: language_service/src/test_utils.rs [hunk 5]
```diff
@@ -10,12 +10,49 @@ use qsc::{
...
     line_column::{Encoding, Position, Range},
     location::Location,
+    packages::prepare_package_store,
     target::Profile,
     LanguageFeatures, PackageStore, PackageType, SourceMap, Span,
 };
 use qsc_project::{PackageGraphSources, PackageInfo};
 use rustc_hash::FxHashMap;
 
+const FAKE_STDLIB_CONTENTS: &str = r#"
```
Comment (idavis) [inline language_service/src/test_utils.rs:20]: Having std lib in the name is a little consfusing to me. Is this just some lib, or is it supposed to be an alternative std lib?

File: compiler/qsc_doc_gen/src/display.rs [hunk 6]
```diff
@@ -540,15 +540,16 @@ impl<'a> Display for FunctorExpr<'a> {
...
-impl<'a> Display for AstPath<'a> {
+impl<'a> Display for AstPathKind<'a> {
     fn fmt(&self, f: &mut Formatter<'_>) -> Result {
-        match self.path.segments.as_ref() {
-            Some(parts) => write!(f, "{parts}.{}", self.path.name.name),
-            None => write!(f, "{}", self.path.name.name),
+        if let ast::PathKind::Ok(path) = self.path {
+            write!(f, "{}", path.full_name())
+        } else {
+            write!(f, "?")
```
Comment (idavis) [inline compiler/qsc_doc_gen/src/display.rs:552]: Should we delegate this `Display` impl to `IncompletePath` which can render the segments with a `?`

File: language_service/src/references/tests.rs [hunk 7]
```diff
@@ -93,12 +93,12 @@ fn std_callable_ref() {
...
-                            line: 1,
-                            column: 26,
+                            line: 2,
+                            column: 18,
                         },
                         end: Position {
-                            line: 1,
-                            column: 30,
+                            line: 2,
+                            column: 22,
```
Comment (idavis) [inline language_service/src/references/tests.rs:101]: Why are the lines and columns changing in these tests?

File: language_service/src/completion/global_items.rs
Comment (minestarks) [reply language_service/src/completion/global_items.rs:1]: OK, in the interest of unblocking the PR, I'm going to postpone the changes to a new PR. This is good to go :)

Just be aware that **reexports** are known to not be working. i.e. you won't get reexported namespaces/items showing up under the correct namespace for dot completions.

In practice, this only seems to matter when you're trying to do completions under "legacy" namespaces like `Microsoft.Quantum.Diagnostics.|` because those legacy names are actually reexports. Otherwise, I haven't found any Q# code actually using reexports yet.

File: language_service/src/completion/path_context.rs [hunk 8]
```diff
@@ -0,0 +1,251 @@
...
+            visit::walk_path(self, path);
+        }
+    }
+
+    fn visit_path_kind(&mut self, path: &'a qsc::ast::PathKind) {
+        let span = match path {
+            qsc::ast::PathKind::Ok(path) => &path.span,
+            qsc::ast::PathKind::Err(Some(incomplete_path)) => &incomplete_path.span,
+            qsc::ast::PathKind::Err(None) => return,
+        };
```
Comment (idavis) [inline language_service/src/completion/path_context.rs:238]: nit: clean up with use to match rest of file

File: language_service/src/completion/path_context.rs [hunk 9]
```diff
@@ -0,0 +1,251 @@
...
+            self.context = Some(PathKind::Struct);
+        }
+    }
+
+    fn visit_path_kind(&mut self, path: &'a qsc::ast::PathKind) {
+        self.qualifier = match path {
+            qsc::ast::PathKind::Ok(path) => path.segments.as_ref().map(AsRef::as_ref),
+            qsc::ast::PathKind::Err(Some(incomplete_path)) => Some(&incomplete_path.segments),
+            qsc::ast::PathKind::Err(None) => None,
+        };
```
Comment (idavis) [inline language_service/src/completion/path_context.rs:71]: nit: clean up with use to match rest of file

File: language_service/src/completion.rs [hunk 10]
```diff
@@ -1,809 +1,334 @@
...
-        }
-        None => before_offset,
-    }
-    .to_string();
-    if !indent.starts_with('\n') {
-        indent.insert(0, '\n');
+        .relative_sources()
+        .find(|s| s.offset == source.offset)
+        .expect("source should exist in the user source map")
+        .name;
```
Comment (idavis) [inline language_service/src/completion.rs:55]: Would it make sense to expose something like `as_relative_source(source: Source)` rather than iterating and cloning all of the sources in the map?
```
let source_name_relative = compilation
        .user_unit()
        .sources
        .as_relative_source(source).name
```

File: playground/src/main.tsx [hunk 11]
```diff
@@ -248,12 +248,6 @@ async function loaded() {
 function registerMonacoLanguageServiceProviders(
   languageService: ILanguageService,
 ) {
-  monaco.languages.setLanguageConfiguration("qsharp", {
```
Comment (ScottCarda-MS) [inline playground/src/main.tsx:251]: I wasn't expecting this to get dropped. Are we not using this anymore?

**Review by ScottCarda-MS — COMMENTED** [review]
partially reviewed, will continue to review

**Comment by minestarks** [comment]
@idavis found a bug: 

Trying to complete

```qsharp
import QIR.in;
```

causes a panic since `in` is a keyword and that's a parse tree I did not expect to have to handle....

Working on the fix.

File: language_service/src/test_utils.rs (hunk 5 above)
Comment (minestarks) [reply language_service/src/test_utils.rs:20]: Yes, it's supposed to replace the stdlib in most language service tests, since it's more minimal (faster to build) and I can easily add any edge case I want to cover.

File: language_service/src/completion/path_context.rs (hunk 9 above)
Comment (minestarks) [reply language_service/src/completion/path_context.rs:71]: Sad story: there are two different structs named `PathKind` in this file, and the qualifiers here are to disambiguate. Maybe I'll use an alias.

File: compiler/qsc_ast/src/ast.rs [hunk 12]
```diff
@@ -1391,57 +1402,65 @@ impl Display for QubitInitKind {
...
-            id: NodeId::default(),
-            span,
-            segments,
-            name: name.into(),
+impl Display for PathKind {
+    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
+        match self {
+            PathKind::Ok(path) => write!(f, "{path}")?,
+            PathKind::Err(Some(incomplete_path)) => {
+                let mut indent = set_indentation(indented(f), 0);
```
[the remark this answers is not in the document]
Comment (minestarks) [reply compiler/qsc_ast/src/ast.rs:1437]: Sorry, I don't understand. What's a normal indentation?

**Comment by minestarks** [comment]
> @idavis found a bug:
> 
> Trying to complete
> 
> ```qsharp
> import QIR.in;
> ```
> 
> causes a panic since `in` is a keyword and that's a parse tree I did not expect to have to handle....
> 
> Working on the fix.

Fixed with my latest commit.

File: playground/src/main.tsx (hunk 11 above)
Comment (minestarks) [reply playground/src/main.tsx:251]: Nope! Good riddance. Called out in the PR description:

> removed wonky word pattern in playground and VS Code because we don't need it anymore now that we have proper `.` completions! Also update trigger characters.

File: compiler/qsc_ast/src/ast.rs (hunk 12 above)
[the remark this answers is not in the document]
Comment (ScottCarda-MS) [reply compiler/qsc_ast/src/ast.rs:1441]: My bad, I misread what this was doing. This seems like the usual way of doing indentation.

File: compiler/qsc_frontend/src/typeck/tests.rs [hunk 13]
```diff
@@ -3080,15 +3080,15 @@ fn local_open() {
             namespace B { function Bar() : () {} }
         "},
         "",
-        &expect![[r#"
+        &expect![[r##"
```
[the remark this answers is not in the document]
Comment (minestarks) [reply compiler/qsc_frontend/src/typeck/tests.rs:3083]: OMG, I took so many passes over this PR trying to revert the sneaky hash changes. I _agree_ with you, but in this file, the warning is suppressed (with `#![allow(clippy::needless_raw_string_hashes)]`) , and the existing test cases are full of `##`. So I thought I should stick to that convention. Now what?! 😆

File: compiler/qsc_ast/src/mut_visit.rs [hunk 14]
```diff
@@ -384,12 +389,26 @@ pub fn walk_path(vis: &mut impl MutVisitor, path: &mut Path) {
...
+    }
+}
+
 pub fn walk_ident(vis: &mut impl MutVisitor, ident: &mut Ident) {
     vis.visit_span(&mut ident.span);
 }
 
-pub fn walk_idents(vis: &mut impl MutVisitor, ident: &mut crate::ast::Idents) {
-    for ref mut ident in &mut *ident.0 {
+pub fn walk_idents(vis: &mut impl MutVisitor, idents: &mut [Ident]) {
```
[the remark this answers is not in the document]
Comment (minestarks) [reply compiler/qsc_ast/src/mut_visit.rs:410]: Great question, I went back and forth on this for a while. Ultimately, I decided to keep `visit_idents` to avoid updating the current usages and bloating this PR even further. 

This came up in a conversation with Ian as well, and my response was basically:

> In the AST, many of choices came down to "how do I avoid touching a billion lines of code", and this was one of them 

I think I'll sleep on it and see on Monday if I want to open a separate refactoring PR, or push even more changes to this one...

File: compiler/qsc_parse/src/completion/collector.rs [hunk 15]
```diff
@@ -0,0 +1,186 @@
...
+    pub fn expect(&mut self, expected: WordKinds) {
+        match self.state {
+            State::AtCursor => self.collected.extend(expected),
+            State::BeforeCursor | State::End => {}
+        }
+    }
+
+    /// The parser has advanced. Update state.
+    pub fn did_advance(&mut self, next_token: &mut Token, scanner_offset: u32) {
+        match self.state {
```
[the remark this answers is not in the document]
Comment (minestarks) [reply compiler/qsc_parse/src/completion/collector.rs:102]: This is semantically an "event handler", as in, it's informing the callee, that advancing _did_ occur. 

I got used to this naming convention in LSP / VS Code APIS (e.g. `didDeleteFile` ) but I guess it's not coming across :D 

Any other suggestions? `on_advance()`? `on_advanced()`? `advanced()` ?

File: language_service/src/completion.rs (hunk 10 above)
Comment (minestarks) [reply language_service/src/completion.rs:55]: Yes, I like that idea.

File: language_service/src/references/tests.rs (hunk 7 above)
Comment (minestarks) [reply language_service/src/references/tests.rs:101]: Because of my changes in `test_utils.rs` - I think I changed some whitespace in the "Fake stdlib" while moving strings around.

File: language_service/src/completion.rs (hunk 2 above)
Comment (minestarks) [reply language_service/src/completion.rs:111]: This turned out to be a preexisting bug. I'll investigate.

Comment (minestarks) [reply language_service/src/completion.rs:111]: Completions were a distraction, actually - having a broken expression at the very end of the cell breaks the parsing of the entire cell, there's no AST at all! Notice none of the editor features (hover, go to def etc) work for _any_ code above that expression.

I'll just file a separate bug for this.

Comment (minestarks) [reply language_service/src/completion.rs:111]: Tracked in #1960

File: compiler/qsc_doc_gen/src/display.rs (hunk 6 above)
Comment (minestarks) [reply compiler/qsc_doc_gen/src/display.rs:552]: I don't think this would work in this case. The AST structs' `Display` impls all print in that "AST-ish" format, like:

```
Err IncompletePath [157-166]:
    Ident 13 [157-160] "Foo"
    Ident 14 [161-164] "Bar"
```

like what you see in the "AST" tab in the playground. These are not meant to be used when pretty printing to the user. (This line actually had a bug where it was exposing the AST-ish format to the user, which I fixed.

We've had to differentiate these impls from the user visible "pretty printing" impls, which live here. That's why there are a bunch of structs in this file that wrap the AST/HIR structs only to implement `Display`.

File: language_service/src/completion.rs (hunk 2 above)
Comment (swernli) [reply language_service/src/completion.rs:111]: Oh wow, I didn't realize that. Thanks for investigating!

File: language_service/src/completion/tests.rs [hunk 16]
```diff
@@ -477,101 +621,75 @@ fn in_block_with_alias() {
...
+            Alias.↘
         }
-        export Bar;
-    }
-    namespace Other {
-        operation Foo() : Unit {}
-        export Foo;
     }"#},
-        &["Foo"],
+        &["Fake", "Alias.Fake", "Library", "Alias.Library"],
```
Comment (ScottCarda-MS) [inline language_service/src/completion/tests.rs:633]: Might be a bit safer to have it also check against function names that aren't in Alias, like `Foo`, and names that are only in Alias but not elsewhere. In the code as-written, it is a bit ambiguous to me if it is validating the `Fake` or the `Alias.Fake`.

File: compiler/qsc_frontend/src/typeck/tests.rs (hunk 13 above)
[the remark this answers is not in the document]
Comment (idavis) [reply compiler/qsc_frontend/src/typeck/tests.rs:3083]: @swernli Can you help out here?

[the remark this answers is not in the document]
Comment (ScottCarda-MS) [reply compiler/qsc_frontend/src/typeck/tests.rs:3083]: In the past, I've had to do a search-and-replace over "##".

**Comment by minestarks** [comment]
> Fixed with my latest commit.

Okay, I regressed things pretty badly with this fix, as @ScottCarda-MS discovered through testing (thanks!). 

The following piece of code should produce a mostly valid AST (save for the error on line 3). 

```qsharp
namespace Sample {
    operation BellStates() : (Result, Result)[] {
        Std.
        let bellStateTuples = [];

        for (label, prepare) in bellStateTuples {}
        return measurements;
    }

    /// doc comment
    operation PreparePhiPlus() : Unit {
    }
}
```

With my latest commit, there's no AST at all.

cc reviewers @idavis  @swernli . I'm going to have to go back to the drawing board with the "keyword after dot" fix that handles `QIR.in` and similar paths. Expect more parser changes :/

File: compiler/qsc_frontend/src/typeck/tests.rs (hunk 13 above)
[the remark this answers is not in the document]
Comment (minestarks) [reply compiler/qsc_frontend/src/typeck/tests.rs:3083]: OK, I just reverted this line to the way it was.

File: language_service/src/tests.rs (hunk 3 above)
Comment (minestarks) [reply language_service/src/tests.rs]: Not exactly, since the contents of the list don't matter for the purposes of this test. But your comment was the push I needed to change this test into something a bit smarter instead of this opaque number that changes every time we touch something.

**Comment by minestarks** [comment]
I pushed what I think is my final update. The parser change to fix the last issue ended up being minimal!

Please sign off if no more concerns.

**Review by ScottCarda-MS — APPROVED** [review]
Discussed offline that the merge conflict is minor. Everything else looks good, and I tested out the feature manually.

File: vscode/src/completion.ts [hunk 17]
```diff
@@ -82,6 +82,8 @@ class QSharpCompletionItemProvider implements vscode.CompletionItemProvider {
       });
       return item;
     });
-    return results.concat(this.samples);
+
+    // Include the samples list for empty documents
+    return document.lineCount === 0 ? results.concat(this.samples) : results;
```
[the remark this answers is not in the document]
Comment (minestarks) [reply vscode/src/completion.ts]: Done.

File: playground/src/main.tsx (hunk 4 above)
Comment (minestarks) [reply playground/src/main.tsx:308]: Done.
